### PR TITLE
[RFR][1LP] cfme/infrastructure/host.py widgetastic conversion

### PR DESCRIPTION
**Pull request by quarckster**
Intent
=================

RHCFQE-2985 cfme/infrastructure/host.py widgetastic conversion
Failures in PRT are not connected with the PR.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit ba8de596c2e6d20045377d2d726a801369dd8985:

`cfme/infrastructure/host.py`:
- :red_circle: Line 88:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 564:12: **F821** *undefined name 'InfraHostDiscoverView'*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/common/host_views.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+class HostDetailsToolbar(View):
+    """Represents host toolbar and its controls."""
+    monitoring = Dropdown(text="Monitoring")
+    configuration = Dropdown(text="Configuration")
+    policy = Dropdown(text="Policy")
+    power = Dropdown(text="Power")
+
+
+class HostDetailsSummaryView(View):
```
Comment (mshriver) [inline cfme/common/host_views.py]: **Required**

I'd like to keep these class names consistent, we've been using 'Entities' for this section of the page in other modules. So `HostDetailsEntities` here, and then in `HostDetailsView` below there should be an `entities` attribute instead of 'contents'

Reason being so that we can iterate over a set of views and consistently validate against a given attribute (toolbar, sidebar, entities, paginator)

Comment (mfalesni) [inline cfme/common/host_views.py]: @mshriver We should probably put together some guidelines name conventions.

File: cfme/common/host_views.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    def in_compute_infrastructure_hosts(self):
+        return (
+            self.logged_in_as_current_user and
+            self.navigation.currently_selected == ["Compute", "Infrastructure", "Hosts"]
+        )
+
+
+class HostQuadIconItem(BaseQuadIconItem):
+
+    @property
```
Comment (izapolsk) [inline cfme/common/host_views.py]: REQUIRED: I don't think we need to create property for any thing there. So, I created data property for all such things.
Please use the same approach.
https://github.com/ManageIQ/integration_tests/blob/master/widgetastic_manageiq/__init__.py#L1829

File: cfme/common/provider_views.py [hunk 3]
```diff
@@ -19,7 +19,16 @@
...
                                   DynaTree,
                                   BootstrapTreeview,
-                                  ProviderItem)
+                                  BaseItem,
+                                  ProviderQuadIconItem,
+                                  ProviderListItem,
+                                  ProviderTileIconItem)
+
+
+class ProviderItem(BaseItem):
```
Comment (izapolsk) [inline cfme/common/provider_views.py]: great, that's what I had planned to do.

File: cfme/common/host_views.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+class HostTileIconItem(BaseTileIconItem):
+    quad_icon = ParametrizedView.nested(HostQuadIconItem)
+
+
+class HostListItem(BaseListItem):
+    pass
+
+
+class HostItem(BaseItem):
+    quad_item = HostQuadIconItem
```
Comment (izapolsk) [inline cfme/common/host_views.py]: :+1:  it is maybe a bit unobvious but I like it :)

File: cfme/common/host_views.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    """Represents hosts toolbar and its controls."""
+    configuration = Dropdown(text="Configuration")
+    policy = Dropdown(text="Policy")
+    lifecycle = Dropdown(text="Lifecycle")
+    monitoring = Dropdown(text="Monitoring")
+    power = Dropdown(text="Power")
+    view_selector = View.nested(ItemsToolBarViewSelector)
+
+
+class HostItems(View):
```
Comment (izapolsk) [inline cfme/common/host_views.py]: we probably need to use Entities in this case. This is my bug. I lost one of renaming commits during one of rebases :(

File: cfme/common/host_views.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    """Represents the view with different items like hosts."""
+    search = View.nested(Search)
+    _quadicons = './/tr[./td/div[@class="quadicon"]]/following-sibling::tr/td/a'
```
Comment (izapolsk) [inline cfme/common/host_views.py]: it seems we can get rid of all these hacks by replacing them with js calls

File: cfme/infrastructure/host.py [hunk 7]
```diff
@@ -173,70 +81,129 @@ def __init__(self, **kwargs):
...
+        view.fill({
+            "name": self.name,
+            "hostname": self.hostname or self.ip_address,
+            "host_platform": self.host_platform,
+            "custom_ident": self.custom_ident,
+            "ipmi_address": self.ipmi_address,
+            "mac_address": self.mac_address
+        })
+        if self.credentials is not None:
+            view.endpoints.default.fill({
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: every credential has view_value_mapping for fill. It would be better to use it here.

File: cfme/common/host_views.py (hunk 2 above)
Comment (quarckster) [inline cfme/common/host_views.py]: @izapolsk  I made it for backward compatibility. Check this for example https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/infrastructure/test_advanced_search_host.py#L28

File: cfme/infrastructure/host.py [hunk 8]
```diff
@@ -173,70 +81,129 @@ def __init__(self, **kwargs):
...
+                "username": self.credentials.principal,
+                "password": self.credentials.secret,
+                "confirm_password": self.credentials.verify_secret,
+            })
+            if validate_credentials:
+                view.endpoints.default.validate_button.click()
+        if self.ipmi_credentials is not None:
+            view.endpoints.fill({
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: view.endpoints.ipmi.fill ?

Comment (quarckster) [inline cfme/infrastructure/host.py]: Of course, will fix that.

File: cfme/common/host_views.py (hunk 6 above)
Comment (quarckster) [inline cfme/common/host_views.py]: @izapolsk  Also I think we should make a base class for items and put it in widgetastic_manageiq.

File: cfme/common/host_views.py (hunk 2 above)
Comment (izapolsk) [inline cfme/common/host_views.py]: I see. But It makes sense to follow the same approach everywhere. 
I had to create data property there because f.e. list items have much more data than quad icons and we also need to make that data accessable. So, I made one dict property for all cases

File: cfme/infrastructure/host.py [hunk 9]
```diff
@@ -173,70 +81,129 @@ def __init__(self, **kwargs):
...
 
-        navigate_to(self, 'Edit')
-        change_stored_password()
-        fill(credential_form, updates.get('credentials'), validate=validate_credentials)
-
-        logger.debug("Trying to save update for host with id: " + str(self.get_db_id))
-        self._submit(cancel, self.forced_saved)
+        view = navigate_to(self, "Edit")
+        changed = view.fill(updates)
+        credentials = updates.get("credentials")
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: you need to pop both credentials before changed. Otherwise, you will get error or warning. I encountered this issue in provider.update

File: cfme/common/host_views.py (hunk 2 above)
Comment (quarckster) [inline cfme/common/host_views.py]: @izapolsk Maybe it make sense to implement `__getattr__()` in that case?

File: cfme/common/host_views.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            password = Input(name="ipmi_password")
+            confirm_password = Input(name="ipmi_verify")
+            validate_button = Button("Validate")
+
+    cancel_button = Button("Cancel")
+
+
+class HostAddView(HostFormView):
+    host_platform = BootstrapSelect("user_assigned_os")
+    add_button = Button("Add")
```
Comment (izapolsk) [inline cfme/common/host_views.py]: We don't have naming conventions yet  but add/cancel/save/etc are used everywhere.  It makes sense to follow one approach everywhere.

File: cfme/infrastructure/host.py [hunk 11]
```diff
@@ -247,79 +214,100 @@ def execute_button(self, button_group, button, cancel=True):
...
-        navigate_to(self, 'All')
-        for page in paginator.pages():
-            if sel.is_displayed(Quadicon(self.name, 'host')):
+        """Checks if the host exists in the UI.
+
+        Returns: :py:class:`bool`
+        """
+        view = navigate_to(self, "All")
+        for page in view.paginator.pages():
+            if self.name in [item.name for item in view.items.get_all_items()]:
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: I guess get_item(by_name=self.name)

File: cfme/infrastructure/host.py [hunk 12]
```diff
@@ -562,79 +519,90 @@ def get_from_config(provider_config_name):
...
+        ipmi_credentials=ipmi_credentials
+    )
 
 
 def wait_for_a_host():
-    navigate_to(Host, 'All')
-    logger.info('Waiting for a host to appear...')
-    wait_for(paginator.rec_total, fail_condition=None, message="Wait for any host to appear",
-             num_sec=1000, fail_func=sel.refresh)
+    """Waits for any host to appear in the UI."""
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: as far as I remember @psav, wanted us to get rid of such wait_for methods

File: cfme/infrastructure/host.py [hunk 13]
```diff
@@ -562,79 +519,90 @@ def get_from_config(provider_config_name):
...
+    Args:
+        host_name (str): A host name as displayed at the quadicon
 
-    if paginator.page_controls_exist():
-        paginator.results_per_page(1000)
-        sel.click(paginator.check_all())
+    Returns: :py:class:`cfme.common.host_views.HostQuadIconItem` instance
+    """
+    view = navigate_to(Host, "All")
+    for page in view.paginator.pages():
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: I guess it makes sense to update get_item in order to stop when item is found

File: cfme/common/host_views.py (hunk 6 above)
Comment (izapolsk) [inline cfme/common/host_views.py]: agree, I also thought of this.

File: cfme/infrastructure/host.py (hunk 12 above)
Comment (quarckster) [inline cfme/infrastructure/host.py]: I believe it should be done in a separate PR. That changes is not related to widgetastic conversion.

**Comment by izapolsk** [comment]
please rebase PR

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7ccef2aa98b2349d59d540358fe011de8007f904:

`cfme/cloud/stack.py`:
- :bangbang: Line 102:13: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 102:101: **E501** *line too long (114 > 100 characters)*

`cfme/fixtures/tag.py`:
- :bangbang: Line 93:1: **E265** *block comment should start with '# '*
- :bangbang: Line 94:1: **E302** *expected 2 blank lines, found 1*
- :heavy_exclamation_mark: Line 118:34: **W292** *no newline at end of file*

`cfme/infrastructure/provider/scvmm.py`:
- :bangbang: Line 12:50: **E231** *missing whitespace after ','*

`cfme/infrastructure/virtual_machines.py`:
- :red_circle: Line 15:1: **N811** *constant imported as non constant*

`cfme/tests/cloud/test_cloud_object_tag_visibility.py`:
- :heavy_exclamation_mark: Line 100:48: **W292** *no newline at end of file*

`cfme/tests/infrastructure/test_infra_object_tag_visibility.py`:
- :heavy_exclamation_mark: Line 68:58: **W292** *no newline at end of file*

`scripts/repo_gen.py`:
- :grey_exclamation: Line 44:9: **T001** *print statement found.*

`widgetastic_manageiq/__init__.py`:
- :bangbang: Line 1896:20: **E126** *continuation line over-indented for hanging indent*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/common/host_views.py (hunk 2 above)
Comment (izapolsk) [inline cfme/common/host_views.py]: okay, and could you add todo to move this to data property later ?

File: cfme/common/host_views.py [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+            self.history_table.is_displayed
+        )
+
+
+class HostDriftAnalysis(ComputeInfrastructureHostsView):
+    apply_button = Button("Apply")
+    drift_sections = CheckableBootstrapTreeview(tree_id="all_sectionsbox")
+
+    @ParametrizedView.nested
+    class drift_analysis(ParametrizedView):  # noqa
```
Comment (izapolsk) [inline cfme/common/host_views.py]: :+1:

File: cfme/common/host_views.py (hunk 6 above)
Comment (izapolsk) [inline cfme/common/host_views.py]: @quarckster, I've extracted BaseEntitiesView. So, you can re-use it here 
https://github.com/ManageIQ/integration_tests/blob/master/widgetastic_manageiq/__init__.py#L2118

Btw, Example of usage - https://github.com/ManageIQ/integration_tests/blob/master/cfme/common/provider_views.py#L298

File: cfme/common/host_views.py [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+            raise ManyItemsFound("Several items with {name} were found".format(name=by_name))
+        return items[0]
+
+
+class HostSideBar(View):
+    """Represents left side bar. It usually contains navigation, filters, etc."""
+    pass
+
+
+class HostsView(ComputeInfrastructureHostsView):
```
Comment (izapolsk) [inline cfme/common/host_views.py]: Please take a look
https://github.com/ManageIQ/integration_tests/blob/master/cfme/common/provider_views.py#L298

File: cfme/common/host_views.py (hunk 10 above)
Comment (izapolsk) [inline cfme/common/host_views.py]: btw, you use both cancel_button and cancel in one PR. I guess we need to stop on one naming approach.

File: cfme/infrastructure/host.py [hunk 16]
```diff
@@ -298,79 +214,101 @@ def execute_button(self, button_group, button, cancel=True):
...
+        view = navigate_to(self, "All")
+        item = view.items.get_item(by_name=self.name)
+        return item.creds.strip().lower() == "checkmark"
 
     def get_datastores(self):
-        """ Gets list of all datastores used by this host"""
-        navigate_to(self, 'Details')
-        list_acc.select('Relationships', 'Datastores', by_title=False, partial=True)
+        """Gets list of all datastores used by this host."""
+        # TODO Refactor this when Datastores will be converted to widgetastic:
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: datastores are converted to widgetastic

File: cfme/infrastructure/host.py [hunk 17]
```diff
@@ -497,111 +422,93 @@ def untag(self, tag):
...
-            self.prerequisite_view.navigation.select('Compute', 'Infrastructure', 'Nodes')
+            self.prerequisite_view.navigation.select("Compute", "Infrastructure", "Nodes")
 
     def resetter(self):
-        tb.select("Grid View")
-        sel.check(paginator.check_all())
-        sel.uncheck(paginator.check_all())
+        if self.view.toolbar.view_selector.selected != "Grid View":
+            self.view.toolbar.view_selector.select("Grid View")
+        self.view.paginator.check_all()
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: please add "if paginator.exists".
Btw, paginator was moved to entities view

File: cfme/infrastructure/host.py [hunk 18]
```diff
@@ -497,111 +422,93 @@ def untag(self, tag):
...
-    prerequisite = NavigateToSibling('All')
+    VIEW = HostDetailsView
+    prerequisite = NavigateToSibling("All")
 
     def step(self):
-        sel.click(Quadicon(self.obj.name, self.obj.quad_name))
-
-    def am_i_here(self):
-        return match_page(summary="{} (Summary)".format(self.obj.name))
+        self.prerequisite_view.items.get_item(by_name=self.obj.name).click()
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: it may not be on current page. you need surf_pages=True.
btw, I added special method for such cases - https://github.com/ManageIQ/integration_tests/blob/master/widgetastic_manageiq/__init__.py#L2101

File: cfme/infrastructure/host.py (hunk 17 above)
Comment (izapolsk) [inline cfme/infrastructure/host.py]: and I'm not sure we need to switch view to Grid now. Since entity object doesn't depend on currently chosen view now.

File: cfme/infrastructure/host.py [hunk 19]
```diff
@@ -613,79 +497,89 @@ def get_from_config(provider_config_name):
...
-    if not do_not_navigate:
-        navigate_to(Host, 'All')
-    for page in paginator.pages():
-        quadicon = Quadicon(host, "host")
-        if sel.is_displayed(quadicon):
-            return quadicon
-    else:
-        raise HostNotFound("Host '{}' not found in UI!".format(host))
+    view = navigate_to(Host, "All")
+    return [entity.name for entity in view.entitis.get_all(surf_pages=True)]
```
Comment (izapolsk) [inline cfme/infrastructure/host.py]: typo

**Comment by quarckster** [comment]
@izapolsk one test failed because of old `cfme.web_ui` element error.
